src/probability/bell_inequality.lean

```diff
@@ -0,0 +1,98 @@
+/-
+Copyright (c) 2022 Ian Jauslin and Alex Kontorovich. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ian Jauslin, Alex Kontorovich
+-/
+import measure_theory.measure.probability_measure
+
+/-!
+# Bell's Inequality
+
+This file proves Bell's Inequality as a probabilistic statement in several forms.
+
+Bell's inequality is the cornerstone of Bell's theorem, which states that quantum mechanics is a
+non-local theory. The proof of Bell's theorem, established by J.S. Bell in 1964, first uses the
+Einstein-Podolsky-Rosen argument to establish that either quantum mechanics is non-local, or all
+quantum observables simultaneously have values (in other words, quantum mechanics is a
+non-contextual hidden variable theory). Bell's inequality is a necessary condition for all the
+observables in an abstract probabilistic theory to simultaneously have values. By showing that, in
+quantum mechanics, Bell's inequality is violated, Bell proved that quantum mechanics is non-local.
+
+The violation of Bell's inequality has also been verified experimentally in many different setups.
+
+## Bell's 1964 inequality
+
+We first prove Bell's original statement of the inequality, which was published in 1964, and will
+thus be called `bell_inequality_1964` in this file.
+
+Consider two observers, $$A$$ and $$B$$, performing measurements that take values in $${-1, 1}$$.
+
+Let `ℙ` be a probability measure, and let `A i` and `B i` for $$i ∈ {0, 1, 2}$$ be random
+variables taking values in $${-1, 1}$$. For convenience, we identify $${-1, 1}$$ with `ℤˣ`. `A i`
+represent the outcomes of measurements done by $$A$$, and `B i` those done by $$B$$. We assume
+perfect anticorrelation between the outcomes of $$A$$ and $$B$$: 𝔼[(A i) (B i)] = -1. Bell's
+inequality states that `𝔼[(A 1) (B 2)] - 𝔼[(A 1) (B 2)] ≤ 1 + 𝔼[(A 2) (B 0)]`.
+
+## TODO
+
+J.S. Bell generalized the inequality in 1975 to include more probabilistic theories. In particular,
+the 1975 statement does not require observables to simultaneously have values. Instead, it is solely
+based on the requirement of locality. The 1975 inequality thus gives a more direct proof of Bell's
+theorem.
+
+## References
+
+* [J.S. Bell, *On the Einstein Podolsky Rosen Paradox*, 1964][MR3790629]
+* [J.S. Bell, *The theory of local beables*, 1975,
+  reproduced in chapter 7 of *Speakable and unspeakable in quantum mechanics*][MR915338]
+
+## See also
+
+`CHSH_inequality_of_comm` is a star-algebra version of Bell's inequality.
+-/
+
+open filter measure_theory
+
+variables {Ω : Type*} [measurable_space Ω] {ℙ : measure Ω} [is_probability_measure ℙ] {f g : Ω → ℤˣ}
+  {A B : fin 3 → Ω → ℤˣ}
+
+private lemma norm_aux (a : ℤˣ) : ‖(a : ℝ)‖ ≤ 1 :=
```

Comment:
We in general avoid using `private` but rather putting auxiliary lemmas in a namespace

Reply:
I don't see why we would avoid `private` when explicitly trying to hide lemmas from the public-facing API. This is not API lemmas, nobody wants to see them cluttering their search.

Reply:
It's not going to clutter the search unless you are importing this file and if you are importing this file it is certainly possible that you want to see these lemmas.

Reply:
No, it will permanently clutter [the mathlib docs search](https://leanprover-community.github.io/mathlib_docs).

Possibly a better solution to it all is to put this file under `archive`. What do you think?

Reply:
That's why I'm suggesting to put it in a obscure namespace so it doesn't really show up in the docs search unless you start by typing the namespace.

Reply:
Wait, but that's not how the search works. It matches substrings, not prefixes, so longer names tend to match more things, not less.